jetty-websocket/javax-websocket-client-impl/src/main/java/org/eclipse/jetty/websocket/jsr356/JsrSessionTracker.java

```diff
@@ -19,37 +19,47 @@
 package org.eclipse.jetty.websocket.jsr356;
 
 import java.util.Collections;
+import java.util.HashSet;
 import java.util.Set;
-import java.util.concurrent.CopyOnWriteArraySet;
+import javax.websocket.Session;
 
 import org.eclipse.jetty.util.component.AbstractLifeCycle;
 import org.eclipse.jetty.util.component.LifeCycle;
 
 public class JsrSessionTracker extends AbstractLifeCycle implements JsrSessionListener
 {
-    private CopyOnWriteArraySet<JsrSession> sessions = new CopyOnWriteArraySet<>();
+    private final Set<JsrSession> sessions = new HashSet<>();
```

Comment:
Using a naked `HashSet` is way better than `CopyOnWriteArraySet`.
However, I think you can do even better with `Collections.newSetFromMap(new ConcurrentHashMap<>())`.

In this way, we don't use a lock at all (as CHM will take care of that) and we don't need to do all the work that `CopyOnWriteArraySet` does.

Reply:
This could work, then we would then return a live version of the set on `getSessions()` instead of a copy like we do with this and `CopyOnWriteArraySet`.

I was thinking in the future we might need some sort of concurrent state in the `SessionTracker`, so that when it is being stopped we can make sure no more sessions are added. And for that we would need something like we have here with the synchronized blocks.

Reply:
@lachlan-roberts you can still return a copy of the list of sessions, to avoid that callers have a direct reference to it - we don't want them doing `Session.getSessions().clear()`.

`SessionTracker` is a child bean of `ServerContainer` or `WebSocketClient`, so when the container is stopped, no more sessions should be added and `SessionTracker` could be `clear()`ed, no?

Reply:
@sbordet not really,  in jetty-10 we try shut down the `ServerContainer` using `Graceful` and close all the open sessions, this is before the container has been stopped and we are still accepting new requests. See PR #4931.

Reply:
Don't worry this is an unrelated issue, I will make the changes you suggested an open an issue for this later.